src/mame/barcrest/mpu4_oki_sampled_sound.cpp

```diff
@@ -0,0 +1,199 @@
+
+//O3 -> G1  O1 -> c2 o2 -> c1
+
+This is a bit of a cheat - since we don't clock into the OKI chip directly, we need to
+calculate the oscillation frequency in advance. We're running the timer for interrupt
+purposes, but the frequency calculation is done by plucking the values out as they are written.
+*/
+
+void mpu4_oki_sampled_sound::ic3_write(offs_t offset, uint8_t data)
+{
+	m_ptm_ic3ss->write(offset, data);
+
+	if (offset == 3)
+	{
+		m_t1 = data;
+	}
+	if (offset == 6)
+	{
+		m_t3h = data;
+	}
+	if (offset == 7)
+	{
+		m_t3l = data;
+	}
+
+	float const num = float(clock()) / ((m_t3l + 1) * (m_t3h + 1));
+	float const denom = std::ceil(float(m_t3h * (m_t3l + 1) + 1) / (2 * (m_t1 + 1))); //need to round up, this gives same precision as chip
+
+	int const freq = int(num * denom);
+
+	if (freq)
+	{
+		m_msm6376->set_unscaled_clock(freq);
+	}
+}
+
+uint8_t mpu4_oki_sampled_sound::ic3_read(offs_t offset)
+{
+	return m_ptm_ic3ss->read(offset);
+}
+
+void mpu4_oki_sampled_sound::ic4_write(offs_t offset, uint8_t data)
+{
+	m_pia_ic4ss->write(offset, data);
+}
+
+uint8_t mpu4_oki_sampled_sound::ic4_read(offs_t offset)
+{
+	return m_pia_ic4ss->read(offset);
+}
+
+void mpu4_oki_sampled_sound::pia_gb_porta_w(uint8_t data)
+{
+	LOG_SS(("%s: GAMEBOARD: PIA Port A Set to %2x\n", machine().describe_context(), data));
+	m_msm6376->write(data);
+}
+
+void mpu4_oki_sampled_sound::pia_gb_portb_w(uint8_t data)
+{
+	uint8_t changed = m_expansion_latch^data;
+
+	LOG_SS(("%s: GAMEBOARD: PIA Port B Set to %2x\n", machine().describe_context(), data));
+
+	if (changed & 0x20)
+	{ // digital volume clock line changed
+		if (!(data & 0x20))
+		{ // changed from high to low,
+			if (!(data & 0x10)) // down
+			{
+				if (m_global_volume < 32) m_global_volume++; //steps unknown
+			}
+			else // up
+			{
+				if (m_global_volume > 0) m_global_volume--;
+			}
+
+			LOG_SS(("%s: GAMEBOARD: Volume Set to %2x\n", machine().describe_context(), data));
+			float percent = (32-m_global_volume)/32.0;
+			m_msm6376->set_output_gain(0, percent);
+			m_msm6376->set_output_gain(1, percent);
+		}
+	}
+	m_msm6376->ch2_w(data & 0x02);
+	m_msm6376->st_w(data & 0x01);
+}
+
+uint8_t mpu4_oki_sampled_sound::pia_gb_portb_r()
+{
+	LOG_SS(("%s: GAMEBOARD: PIA Read of Port B\n", machine().describe_context()));
+	uint8_t data = 0;
+	// b7 NAR - we can load another address into Channel 1
+	// b6, 1 = OKI ready, 0 = OKI busy
+	// b5, vol clock
+	// b4, 1 = Vol down, 0 = Vol up
+	//
+
+	if (m_msm6376->nar_r()) data |= 0x80;
+	else                    data &= ~0x80;
+
+	if (m_msm6376->busy_r()) data |= 0x40;
+	else                     data &= ~0x40;
+
+	return data | m_expansion_latch;
+}
+
+WRITE_LINE_MEMBER(mpu4_oki_sampled_sound::pia_gb_ca2_w)
+{
+	LOG_SS(("%s: GAMEBOARD: OKI RESET data = %02X\n", machine().describe_context(), state));
+	if ((m_last_reset != state) && !state)
+	{
+		m_msm6376->reset();
+	}
+//  reset line
+}
```

Comment:
Should this be updating `m_last_reset`?  It looks like it’s supposed to detect falling edges.